### PR TITLE
c code of maxpool

### PR DESCRIPTION
**Pull request by yaoli**
Help needed in C code. 
In line 386, z[0] is always 0 no matter the value of the var 'maximum'.
I don't understand why.

### CODE REVIEW

File: theano/tensor/signal/downsample.py [hunk 1]
```diff
@@ -268,78 +268,123 @@ def grad(self, inp, grads):
...
+                          (dtype_%(z)s*)(PyArray_GETPTR4(%(z)s, b, k, i, j)));
+                    // use the first element  as the initial value of maximum
+                    maximum = ((dtype_%(x)s*)(PyArray_GETPTR4(%(x)s,b,k,r_st,c_st)))[0];
+                    printf("maximum= %%.5f ",maximum);
+                    // go through the pooled region in the input
+                    for(int m=r_st; m<r_end; m++)
+                    {
+                      for(int n=c_st; n<c_end; n++)
+                      {
+                        dtype_%(x)s a = ((dtype_%(x)s*)(PyArray_GETPTR4(%(x)s,b,k,m,n)))[0];
```
Comment (lamblin) [inline theano/tensor/signal/downsample.py]: Here, `r_st` and `r_end` are indices into the _padded_ image, so are `m` and `n` as well.
So, you cannot call `PyArray_GETPTR4` on the _original_ image with those indices.
Either you would have to make `r_st` (and so on) indices into the original image, or to use `m - %(pd0)s` and `n - %(pd1)s`.

File: theano/tensor/signal/downsample.py [hunk 2]
```diff
@@ -268,78 +268,123 @@ def grad(self, inp, grads):
...
+        c = PyArray_DIMS(%(x)s)[3];
+        r += %(pd0)s * 2;
+        c += %(pd1)s * 2;
+        
         if (%(ignore_border)s)
         {
-            x_shp0_usable = z_shp0 * %(ds0)s;
-            x_shp1_usable = z_shp1 * %(ds1)s;
+            z_r = (r - %(ds0)s) / %(st0)s + 1;
+            z_c = (c - %(ds1)s) / %(st1)s + 1;
```
Comment (lamblin) [inline theano/tensor/signal/downsample.py]: You should raise an exception if `ignore_border` is true but `padding` is not `(0, 0)`.

File: theano/tensor/signal/downsample.py (hunk 1 above)
Comment (yaoli) [inline theano/tensor/signal/downsample.py]: Are they? I turned them into coordinates in the original images before going into the loop on them, for example, at line 364, 365

Comment (lamblin) [inline theano/tensor/signal/downsample.py]: I don't understand what you mean. For the first patch (pooling region), for instance, `i = 0`, `j = 0`, so `r_st = 0` and `c_st = 0`. The first pooling region starts at the corner of the padded image, so these indices refer to the corner of the padded image.
Then, you take the maximum of `c_st` and `%(pd1)s`, to have the effective start point, which is the coordinate of the original image inside the padded one. Until now it is consistent: you are using coordinates where 0 mean the start of the padded image, and the first pixel is at `(%(pd0)s, %(pd1)s)`.
The problem is that you then try to access `x[b, k, r_st, c_st]` first, which is `x[b, k, pd0, pd1]`, which is not the first pixel of the original image. You want to access `x[b, k, 0, 0]` instead.

**Comment by lamblin** [comment]
Also, you will need to rebase the pull request on top of the latest central/master.

File: theano/tensor/signal/downsample.py (hunk 1 above)
Comment (yaoli) [inline theano/tensor/signal/downsample.py]: Oh, I've fixed that part.

Still not working.

variable m can get negative values, which I don't see why.

File: theano/tensor/signal/downsample.py (hunk 2 above)
Comment (lamblin) [inline theano/tensor/signal/downsample.py]: This should still be addressed.

File: theano/tensor/signal/downsample.py (hunk 1 above)
Comment (lamblin) [inline theano/tensor/signal/downsample.py]: That was done.

**Comment by lamblin** [comment]
Some tests are failing in Travis.

File: theano/tensor/signal/downsample.py [hunk 3]
```diff
@@ -287,77 +287,116 @@ def grad(self, inp, grads):
...
-                    mini_j = ((mini_j + 1) == %(ds1)s) ? 0 : mini_j+1;
-                    zj += (mini_j == 0);
+            for(int b=0; b<PyArray_DIMS(%(x)s)[0]; b++){
+              for(int k=0; k<PyArray_DIMS(%(x)s)[1]; k++){
+                for(int i=0; i< z_r; i++){
+                  r_st = i * %(st0)s;
+                  r_end = r_st + %(ds0)s;
+                  // skip the padding
+                  r_st = r_st < %(pd0)s ? %(pd0)s : r_st;
+                  r_end = r_end > (r - %(pd0)s) ? r - %(pd0)s : r_end;
```
Comment (lamblin) [inline theano/tensor/signal/downsample.py]: If `ignore_border` is True, actually, you have to stop even before.

File: theano/tensor/signal/downsample.py [hunk 4]
```diff
@@ -287,77 +287,116 @@ def grad(self, inp, grads):
...
+                  // from padded_img space to img space
+                  r_st -= %(pd0)s;
+                  r_end -= %(pd0)s;
+                  for(int j=0; j<z_c; j++){
+                    c_st = j * %(st1)s;
+                    c_end = c_st + %(ds1)s;
+                    // skip the padding
+                    c_st = c_st < %(pd1)s ? %(pd1)s : c_st;
+                    c_end = c_end > (c - %(pd1)s) ? c - %(pd1)s : c_end;
```
Comment (lamblin) [inline theano/tensor/signal/downsample.py]: Same here.

**Comment by lamblin** [comment]
This PR is for #2582

**Comment by yaoli** [comment]
I'll have a look at it tomorrow.

**Comment by yaoli** [comment]
now all the tests passed!

File: theano/tensor/signal/downsample.py (hunk 3 above)
Comment (lamblin) [inline theano/tensor/signal/downsample.py]: Done.

File: theano/tensor/signal/downsample.py (hunk 4 above)
Comment (lamblin) [inline theano/tensor/signal/downsample.py]: Done.

**Comment by lamblin** [comment]
Thanks!

**Pull request merged**